ouroboros-network-framework/src/Ouroboros/Network/ConnectionHandler.hs

```diff
@@ -120,29 +120,22 @@ data Handle (muxMode :: Mx.Mode) initiatorCtx responderCtx versionData bytes m a
         hVersionData    :: !versionData
       }
 
-type family MkMuxConnectionHandler (muxMode :: Mx.Mode) socket initiatorCtx responderCtx
-                                   peerAddr versionNumber versionData bytes m a b =
-  result | result -> socket where
-  MkMuxConnectionHandler Mx.InitiatorMode socket initiatorCtx responderCtx peerAddr
-                                          versionNumber versionData bytes m a b =
-    MuxConnectionHandler Mx.InitiatorMode socket initiatorCtx responderCtx peerAddr
-                                          versionNumber versionData bytes m a b
-
-  MkMuxConnectionHandler Mx.ResponderMode socket initiatorCtx responderCtx peerAddr
-                                          versionNumber versionData bytes m a b =
-       (   StrictTVar m (Maybe ResponderCounters)
-        -> Tracer m (WithBearer (ConnectionId peerAddr) Trace))
-    -> MuxConnectionHandler Mx.ResponderMode socket initiatorCtx responderCtx peerAddr
-                                             versionNumber versionData bytes m a b
-
-  MkMuxConnectionHandler Mx.InitiatorResponderMode socket initiatorCtx responderCtx peerAddr
-                                                   versionNumber versionData bytes m a b =
-       (versionData -> DataFlow)
+data MkMuxConnectionHandler (muxMode :: Mx.Mode) socket initiatorCtx responderCtx
+                            peerAddr versionNumber versionData bytes m a b where
+  MuxInitiatorConnectionHandler :: MkMuxConnectionHandler
+                                     Mx.InitiatorMode socket initiatorCtx responderCtx
+                                     peerAddr versionNumber versionData bytes m a b
+  MuxResponderConnectionHandler :: (   StrictTVar m (Maybe ResponderCounters)
+                                    -> Tracer m (WithBearer (ConnectionId peerAddr) Trace))
+                                -> MkMuxConnectionHandler
+                                     Mx.ResponderMode socket initiatorCtx responderCtx
+                                     peerAddr versionNumber versionData bytes m a b
+  MuxInitiatorResponderConnectionHandler
+    :: (versionData -> DataFlow)
```

Comment:
I think this commit looks much better, no need to carry around a singleton when we can do it with pretty idiomatic Haskell! :)